src/core/lib/lattice/hal/default/dcrtpoly.cpp

```diff
@@ -1376,6 +1379,32 @@ std::shared_ptr<typename DCRTPolyImpl<VecType>::Params> DCRTPolyImpl<VecType>::G
     return std::make_shared<DCRTPolyImpl::Params>(2 * this->GetRingDimension(), moduliQP, rootsQP);
 }
 
+template <typename VecType>
+void DCRTPolyImpl<VecType>::TimesQovert(
+    const std::shared_ptr<DCRTPolyImpl::Params> paramsQ,
+    const std::vector<NativeInteger> &tInvModq,
+    const std::vector<NativeInteger> &tInvModqPrecon,
+    const NativeInteger &t,
+    const NativeInteger &NegQModt,
+    const NativeInteger &NegQModtPrecon) {
+  usint sizeQ = m_vectors.size();
+  if (tInvModq.size() < sizeQ || tInvModqPrecon.size() < sizeQ) {
+    OPENFHE_THROW(math_error, "Sizes of vectors do not match.");
+  }
+  usint ringDim = this->GetRingDimension();
+#pragma omp parallel for
+  for (size_t i = 0; i < sizeQ; i++) {
+    const NativeInteger &qi = paramsQ->GetParams()[i]->GetModulus();
+    const NativeInteger &tInvModqi = tInvModq[i];
+    const NativeInteger &tInvModqiPreconi = tInvModqPrecon[i];
+    for (usint ri = 0; ri < ringDim; ri++) {
```

Comment:
I spent a few hours trying to move the multiplication by -Q (mod t) into the encoding, but it turns out to be a very big change because there are certain places where a plaintext needs to be scaled up by Q/t (addition), and there are certain places where it should not be (EvalMult ct by pt and EvalMerge). Putting it in the encoding does not make sense in general, especially if we add more cases in the future where the plaintext should not be scaled up by Q/t. Also, even in the encoding, we multiply each element of the polynomial by Q/t separately, so I'd rather just leave this here for now. I changed tInv (mod q) to be a vector-level operation though.